File: lab20_categorical_encoding.py
# Title     : Categorical Encoding no Python
# Objective : Conhecer algumas ferramentas que a linguagem Python tem a nos oferecer para trabalharmos com categorical encoding -> transformar dados categoricos em dados numéricos
# Created by: accol
# Created on: 08/06/2020


import pandas as pd 
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.compose import make_column_transformer


# Para esse laboratório trabalharemos com o conjunto de dados Credit.csv
# Inicalmente vamos visualizar os dados sem qualquer transformação
dataset = pd.read_csv('D:/Users/Angelo/AULAS/Dados/Credit.csv')
print('\nObservando nosso conjunto de Dados sem qualquer transformação')
print('\n', dataset.head())
print('\nA dimensão da nossa base é: ', dataset.shape)

# Faremos os nossos testes com duas variáveis categóricas => personal_status e other_parties -> índices 8 e 9 de nossa base. Criaremos uma variável X que receberá estas duas colunas -> note o uso do método .values
X = dataset.iloc[:, 8:10].values
print('\nObservando nossa variável X -> antes da transformação')
print('\n', X)
print('\nA dimensão da variável X é: ', X.shape)

# Vamos agora iniciar a transformação da nossa variável X
labelencoder = LabelEncoder()
X[:,0] = labelencoder.fit_transform(X[:, 0])

onehotencoder = make_column_transformer((OneHotEncoder(categories='auto', sparse=False), [1]), remainder='passthrough')
X = onehotencoder.fit_transform(X)
print('\nObservando nossa variável X -> usando OneHotencoder')
print('\n', X)
print('\nA dimensão da variável X é: ', X.shape)
# Só relembrando antes de utilizar esse modelo você deverá excluir uma coluna -> normalmente exclui-se a coluna 0

File: lab20_categorical_encoding.R
# Title     : Categorical Encoding no R
# Objective : Conhecer algumas ferramentas que a linguagem R tem a nos oferecer para trabalharmos com categorical encoding -> transformar dados categoricos em dados numéricos
# Created by: accol
# Created on: 08/06/2020


library(mltools)
library(data.table)


# Para esta prática utilizaremos os dados do Titanic nativos do R

cat('\nVamos observar nosso conjunto de dados\n', Titanic)
cat('\nPara bem tratarmos nossos dados, precisamos conhecer o tipo de dados: ', class(Titanic))
# Observe que nosso conjunto de dados está formato de tabelas -> para usarmos as ferramentas do R, precisaremos antes converter esses dados para o tipo adequado -> no caso vamos transformá-lo em um DataFrame

tit = as.data.frame(Titanic)
cat('\nVamos observar nosso conjunto de dados -> transformado\n')
print(tit)
cat('\nConferindo nossos dados, analisando a conversar de tipo: ', class(tit))

# Primeiramente vamos trabalhar com label Encoding -> vamos trabalhar com a coluna Classe, sexo e idade (categóricas)

labenc = data.matrix(tit[, 1:3])
cat('\nVamos observar nosso conjunto de dados labenc -> transformado\n')
print(labenc)

# Vamos testar agora o ONE-HOT ENCODING para as mesmas três colunas Classe, sexo e idade categóricas
# Para o ONE-HOT ENCODING teremos que transformar novamente os dados em tabela

hotenco = one_hot(as.data.table(tit[, 1:3]))
cat('\nVamos observar nosso conjunto de dados hotenco -> transformado\n')
print(hotenco)

# Para trabalhar com ONE-HOT ENCODING vamos precisar suprimir uma coluna, não se esqueça disso, ok
